organization/ircam_oauth2_provider/views.py

```diff
@@ -0,0 +1,151 @@
+    
+    logger = Logger(logging.DEBUG) if (settings.DEBUG) else Logger(logging.INFO)
+        
+    
+    provider_id = IrcamAuthProvider.id
+    access_token_url = '{}/o/token/'.format(settings.OAUTH_SERVER_BASEURL)  # Called programmatically, must be reachable from container
+#    authorize_url = '{}/o/authorize/'.format(settings.OAUTH_SERVER_BASEURL)  # Accessed by the client so must be host-reachable
+    authorize_url = '{}/o/authorize/'.format(settings.USER_SERVER_BASEURL)  # Accessed by the client so must be host-reachable
+    profile_url = '{}/profile/'.format(settings.OAUTH_SERVER_BASEURL)
+
+    def create_or_update_person(self,user):
+        # TODO Person.object create or update
+        try :
+            # get Person by mail
+            person, created = Person.objects.get_or_create(user_id=user.id)
+            
+            person.user = user
+            # if a Person is created or updated, populate it
+            # if created:
+            person.title=user.first_name+" "+user.last_name
+            person.first_name=user.first_name
+            person.last_name=user.last_name
+            person.email=user.email
+
+            person.save()
+
+        except ObjectDoesNotExist:
+            person = Person.objects.get(
+                user_id=user.id
+            )
+        except IntegrityError:
+            pass
+
+    def create_user_socialaccount(self,request,extra_data):
+
+        user, created = User.objects.update_or_create(
+            username = extra_data['username'],
+            defaults={
+                'first_name': extra_data['first_name'],
+                'last_name': extra_data['last_name'],
+                'is_active': True,
+                'is_superuser': False,
+                'is_staff': False,
+                'email': extra_data['email'],
+                'password': 'unique.password'
+            }
+        )
+        user.save()
+        self.logger.info('Local user {0} id:{1} username:{2} email:{3}'.format('created: ' if created else 'updated: ',user.id,user.username,user.email))
+        
+        # Creating or updating Person
+        self.create_or_update_person(user)
+       
+        # Creating the allauth social account so the user can log in through Ircam Auth
+        social_account = SocialAccount(user=user,
+                                    provider='ircamauth',
+                                    uid=extra_data['id'],
+                                    extra_data={'id': user.id, 'username': user.username, 'email': user.email})
+        social_account.save()
+        self.logger.info('Social account created. User {0} '.format(user.username))
+
+        return self.get_provider().sociallogin_from_response(request, extra_data)
+
+    def update_localuser(self,hello,extra_data):
+        from django.contrib.auth.models import User
+
+        user = User.objects.filter( username__icontains=extra_data['username'] )[0]
```

Comment:
why use `icontains` instead of `contains` ? If some `username` has capital letter that could lead to a vulnerability.

Reply:
done